src/Microsoft.AspNetCore.Routing/DispatcherMiddleware.cs

```diff
@@ -94,8 +94,7 @@ private Task<Matcher> InitializeAsync()
                 null) == null)
             {
                 // This thread won the race, do the initialization.
-                var dataSource = new CompositeEndpointDataSource(_options.Value.DataSources);
-                var matcher = _matcherFactory.CreateMatcher(dataSource);
```

Comment:
I'm not really sure why I didn't do this from the beginning. I had a some concerns about initialization ordering, but it seems like that's not a problem as long as the composite data source stays lazy.